File: viscoelastic_2d_ssg.py
# Решение уравнений сесйсмики в средах с затуханием на смещенной сетке

import numpy as np

import matplotlib.pyplot as plt
import matplotlib.animation as animation

from numba import jit, njit

lamb = 20000 # Pa
mu = 40000 # Pa
rho = 2700 # kg/m^2

tau_p = 0
tau_s = 0
tau = 10
Lx = 10 #meter
Ly = 10 #meter

Nx = 1000
Ny = 1000

Nt = 200
T = 0.2

dt = T / Nt
dx = Lx / Nx
dy = Ly / Ny

micro_coord = (Nx // 2,   Ny // 2 - 20)
source_coord = (Nx // 2, Ny // 2)
alpha = np.sqrt((lamb + 2 * mu) / rho)

stability_cryteria = np.max([dx, dy]) / (np.sqrt(alpha * (27/24 - 1/24)))
assert dt < stability_cryteria, f"Нет устойчивости! {dt} < {stability_cryteria}"


snap_freq = 1
fM = 40 # Herz

@njit(parallel=True, fastmath = True)
def run():

  vx = np.zeros((Nx, Ny))
  vy = np.zeros((Nx, Ny))

  pxx = np.zeros((Nx, Ny))
  pyy = np.zeros((Nx, Ny))
  pxy = np.zeros((Nx, Ny))

  rxx = np.zeros((Nx, Ny))
  ryy = np.zeros((Nx, Ny))
  rxy = np.zeros((Nx, Ny))


  # 1 / dx * (27/24 * () - 1 / 24 * ())
  p1P = (lamb + 2 * mu) * (1 + tau_p)
  p1S = mu * (1 + tau_s)
  p1d = p1P - 2 * p1S

  p2P = (lamb + 2 * mu) * tau_p
  p2S = mu * tau_s
  p2d = p2P - 2 * p2S


  snapshots = np.zeros((Nt // snap_freq, Nx, Ny))

  ricker_wavelet = lambda t: (1 - 2 * np.pi**2 * fM**2 * t**2) * np.exp(- np.pi**2 * fM**2 * t**2)

  micro_data = np.zeros(Nt // snap_freq)

  snap_idx = 0

  for n in range(0, Nt):
    # print(f"process {n} / {Nt}", end="\r")
    # pxx[*source_coord] = dt * ricker_wavelet(n * dt - 0.1)
    # pyy[*source_coord] = ricker_wavelet(n * dt - 0.1)
    # pxy[*source_coord] = ricker_wavelet(n * dt - 0.1)


    # update velocities
    for i in range(5, Nx - 4):
      for j in range(5, Ny - 4):
        pxx_x = 1 / dx * (27/24 * (pxx[i + 1, j] - pxx[i, j]) - 1 / 24 * (pxx[i + 2, j] - pxx[i - 1, j]))
        pxy_y = 1 / dy * (27/24 * (pxy[i, j] - pxy[i, j - 1]) - 1 / 24 * (pxy[i, j + 1] - pxy[i, j - 2]))
        vx[i, j] = vx[i, j] + dt / rho * (pxx_x + pxy_y)

        pyy_y = 1 / dy * (27/24 * (pyy[i, j + 1] - pyy[i, j]) - 1 / 24 * (pyy[i, j + 2] - pyy[i, j - 1]))
        pxy_x = 1 / dx * (27/24 * (pxy[i, j] - pxy[i - 1, j]) - 1 / 24 * (pxy[i + 1, j] - pxy[i - 2, j]))
        vy[i, j] = vy[i, j] + dt / rho * (pxy_x + pyy_y)
    
    #vx[*source_coord] += dt * np.sin(0.15 * n * dt)
    #vy[*source_coord] += dt * np.sin(0.15 * n * dt)

    vx[*source_coord] += dt * ricker_wavelet(n * dt - 0.05)
    vy[*source_coord] += dt * ricker_wavelet(n * dt - 0.05)

    # update stress
    for i in range(5, Nx - 4):
      for j in range(5, Ny - 4):
        vx_x = 1 / dx * (27/24 * (vx[i, j] - vx[i - 1, j]) - 1 / 24 * (vx[i + 1, j] - vx[i - 2, j]))
        vy_y = 1 / dy * (27/24 * (vy[i, j] - vy[i, j - 1]) - 1 / 24 * (vy[i, j + 1] - vy[i, j - 2]))

        pxx[i, j] = pxx[i, j] + dt * (p1P * vx_x + p1d * vy_y) # + 0.5 * rxx[i, j]  # добавим вторую половинку с нового слоя позже
        pyy[i, j] = pyy[i, j] + dt * (p1d * vx_x + p1P * vy_y) # + 0.5 * ryy[i, j]  # добавим вторую половинку с нового слоя позже

        vx_y = 1 / dy * (27/24 * (vx[i, j + 1] - vx[i, j]) - 1 / 24 * (vx[i, j + 2] - vx[i, j - 1]))
        vy_x = 1 / dx * (27/24 * (vy[i + 1, j] - vy[i, j]) - 1 / 24 * (vy[i + 2, j] - vy[i - 1, j]))

        pxy[i, j] = pxy[i, j] + dt * (p1S * vx_y + p1S * vy_x)  #+ 0.5 * rxy[i, j] # добавим вторую половинку с нового слоя позже

    # update memory variables
    # for i in range(5, Nx - 4):
    #   for j in range(5, Ny - 4):
    #     vx_x = 1 / dx * (27/24 * (vx[i, j] - vx[i - 1, j]) - 1 / 24 * (vx[i + 1, j] - vx[i - 2, j]))
    #     vy_y = 1 / dy * (27/24 * (vy[i, j] - vy[i, j - 1]) - 1 / 24 * (vy[i, j + 1] - vy[i, j - 2]))

    #     rxx[i, j] = 1. / (tau + 1) * (- rxx[i, j] - dt * (p2P * vx_x + p2d * vy_y))
    #     ryy[i, j] = 1. / (tau + 1) * (- ryy[i, j] - dt * (p2d * vx_x + p2P * vy_y))

    #     vx_y = 1 / dy * (27/24 * (vx[i, j + 1] - vx[i, j]) - 1 / 24 * (vx[i, j + 2] - vx[i, j - 1]))
    #     vy_x = 1 / dx * (27/24 * (vy[i + 1, j] - vy[i, j]) - 1 / 24 * (vy[i + 2, j] - vy[i - 1, j]))

    #     rxy[i, j] = 1. / (tau + 1) * (- rxy[i, j] - dt * (p2S * vx_y + p2S * vy_x))

    #     pxx[i, j] += 0.5 * rxx[i, j]
    #     pyy[i, j] += 0.5 * ryy[i, j]
    #     pxy[i, j] += 0.5 * rxy[i, j]

    if (n % snap_freq == 0):
      snapshots[snap_idx] = vx
      micro_data[snap_idx] = vx[*micro_coord] #vx[*micro_coord]
      snap_idx += 1
  
  return snapshots, micro_data

snapshots, micro_data = run()

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 8))

frames = []

Umin = np.quantile(snapshots[:, :, :], 0.75)
Umax = np.quantile(snapshots[:, :, :], 0.75)

signal = np.zeros(Nt // snap_freq)
t = np.linspace(0, T, Nt  // snap_freq)

for k in range(Nt  // snap_freq):
    frame = ax1.imshow(snapshots[k, :, :], extent=[0, Lx, 0, Ly])

    frame1, = ax1.plot(micro_coord[0] * dx, micro_coord[1] * dy, 'ro')

    signal, = ax2.plot(t[:k], micro_data[:k], 'b-')

    frames.append([frame, frame1, signal])

ani = animation.ArtistAnimation(fig,frames,interval=10, blit=True, repeat_delay=1000)

#plt.colorbar(frame) 
plt.show()

File: viscoelastic_2d_ssg.jl
using Plots;
using Base.Threads;
using LoopVectorization;
using ProgressBars;

Threads.nthreads() = 8

# wavefield clip
clip = 2.5e-2

# viscoelastic params
τP = 0
τS = 0
τ = 10

#initial velocities
vp = 2955.0
vs = 2362.0
# density
ρ = 7100.0
# lame params
μ = ρ * vs * vs 
λ = ρ * vp * vp - 2 * μ


# Настройки сетки
Lx = 3005 # m
Ly = 3005 # m

Nx = 601 # points
Ny = 601 # points

dx = Lx / Nx # m
dy = Ly / Ny # m

# Настройки временной сетки
T  = 1 # sec
dt = 0.6e-3
Nt = trunc(Int, T / dt)
# Nt = 1000
# dt = T / Nt

# source wavelet params
fc = 17.0
tshift = 0.0
ts = 1.0 / fc

# M = 6 # order of approx
# α = [1.221336364746061, -0.0969314575195135, 0.017447662353510373, -0.0029672895159030804, 0.00035900539822037244, -2.1847811612209015e-5] # spacial finite difference coefficient

M = 2
α = [1.125 -0.041666666666666664]

vx = zeros(Nx, Ny)
vy = zeros(Nx, Ny)

pxx = zeros(Nx, Ny)
pyy = zeros(Nx, Ny)
pxy = zeros(Nx, Ny)

rxx = zeros(Nx, Ny)
ryy = zeros(Nx, Ny)
rxy = zeros(Nx, Ny)

p1P = (λ + 2 * μ) * (1 + τP)
p1S = μ * (1 + τS)
p1d = p1P - 2 * p1S

p2P = (λ + 2 * μ) * τP
p2S = μ * τS
p2d = p2P - 2 * p2S

fm = 40 # for ricker wavelet

ricker_wavelet(t) = (1. - 2 * pi^2 * fm^2 * t^2) * exp(-pi^2 * fm^2 * t^2)

micro_coord = [Nx ÷ 2, Ny ÷ 2]
micro_data = zeros(Nt)

# PML настрйока
Npml = 10 # толщина PML слоя

Npml_y_start = M + 1
Npml_y_end = Ny - M

Nx_start = M + Npml + 1
Nx_end = Nx - M - Npml

Ny_start = M + Npml + 1
Ny_end = Ny - M - Npml


# Perfectly matched layer

κmax = 1

Rc = 0.1 / 100 #процент теоретического отражения
d0 = - 3 * vp * log(Rc) / 2 / Npml

bx = ones(Nx)
ax = ones(Nx)
κx = ones(Nx)

by = ones(Ny)
ay = ones(Ny)
κy = ones(Ny)

αmax = pi * 25

PML_RIGHT_EDGE = true
PML_LEFT_EDGE = true
PML_TOP_EDGE = true
PML_BOTTOM_EDGE = true


for p in 1:Npml
  d = d0 * ((p - 1) / (Npml - 1))^2
  α = αmax - αmax * (p - 1) / (Npml - 1)

  if PML_RIGHT_EDGE
    i = Nx - M - Npml + p

    κx[i] = 1 + (κmax - 1) * ((p - 1) / (Npml - 1))^2
    bx[i] = exp(-(d / κx[p] + α) * dt) 
    ax[i] = d / (κx[p] * (d + κx[p] * α)) * (bx[p] - 1)
  end

  if PML_LEFT_EDGE
    i = M + Npml + 1 - p

    κx[i] = 1 + (κmax - 1) * ((p - 1) / (Npml - 1))^2
    bx[i] = exp(-(d / κ[p] + α) * dt) 
    ax[i] = d / (κ[p] * (d + κ[p] * α)) * (b[p] - 1)
  end

  if PML_TOP_EDGE
    j = Ny - M - Npml + p

    κy[j] = 1 + (κmax - 1) * ((p - 1) / (Npml - 1))^2
    by[j] = exp(-(d / κ[p] + α) * dt) 
    ay[j] = d / (κ[p] * (d + κ[p] * α)) * (b[p] - 1)
  end

  if PML_Bottom_EDGE
    j = M + Npml + 1 - p

    κx[j] = 1 + (κmax - 1) * ((p - 1) / (Npml - 1))^2
    bx[j] = exp(-(d / κ[p] + α) * dt) 
    ax[j] = d / (κ[p] * (d + κ[p] * α)) * (b[p] - 1)
  end
end


ψx_pxx  = zeros(Nx, Ny)
ψx_pxy  = zeros(Nx, Ny)
ψx_vx   = zeros(Nx, Ny)
ψx_vy   = zeros(Nx, Ny)

ψy_pxy  = zeros(Nx, Ny)
ψy_pyy  = zeros(Nx, Ny)
ψy_vy   = zeros(Nx, Ny)
ψy_vx   = zeros(Nx, Ny)

@gif for n in ProgressBar(1:Nt)
  # update particle velocities 
  @tturbo for i in Nx_start:Nx_end
    for j in Ny_start:Ny_end
        pxx_x = 0
        pyy_y = 0
        pxy_x = 0
        pxy_y = 0     

        for m in 1:M
          pxx_x += α[m] * (pxx[i + m, j] - pxx[i- m + 1, j ])
          pyy_y += α[m] * (pyy[i, j + m] - pyy[i , j- m + 1])
          pxy_x += α[m] * (pxy[i + m - 1, j] - pxy[i-m, j])
          pxy_y += α[m] * (pxy[i, j + m - 1] - pxy[i , j- m]) 
        end       
        
        vx[i, j] = vx[i, j] + dt / ρ * (pxx_x / dx + pxy_y / dy)
        vy[i, j] = vy[i, j] + dt / ρ * (pxy_x / dx + pyy_y / dy)
    end 
  end 

  # add source wavelet
  vy[Nx ÷ 2, Ny ÷ 2] = vy[Nx ÷ 2, Ny ÷ 2] + ricker_wavelet(n * dt - 0.1)  

  # update stresses
  @tturbo for i in Nx_start:Nx_end
    for j in Ny_start:Ny_end
        vxx = 0
        vyy = 0        
        vyx = 0
        vxy = 0

        for m in 1:M
          vxx += α[m] * (vx[i+ m - 1, j ] - vx[i-m, j])
          vyy += α[m] * (vy[i, j+ m - 1 ] - vy[i, j-m])  
          vyx += α[m] * (vy[i+m, j] - vy[i- m + 1, j ])
          vxy += α[m] * (vx[i , j+m] - vx[i , j- m + 1])
        end    

        pxx[i, j] = pxx[i, j] + dt  * (  p1P * vxx / dx + p1d * vyy / dy + 0.5 * rxx[i, j])
        pyy[i, j] = pyy[i, j] + dt  * (  p1d * vxx / dx + p1P * vyy / dy + 0.5 * ryy[i, j])
        pxy[i, j] = pxy[i, j] + dt  * (  p1S * vyx / dx + p1S * vxy / dy + 0.5 * rxy[i, j])
    end
  end

  # update memory variables
  @tturbo for i in Nx_start:Nx_end
    for j in Ny_start:Ny_end
        vxx = 0
        vyy = 0        
        vyx = 0
        vxy = 0

        for m in 1:M
          vxx += α[m] * (vx[i + m - 1, j] - vx[i-m, j])
          vyy += α[m] * (vy[i, j + m - 1] - vy[i, j-m])  
          vyx += α[m] * (vy[i+m, j] - vy[i - m + 1, j ])
          vxy += α[m] * (vx[i , j+m] - vx[i , j - m + 1])
        end    
         
        # update stresses
        rxx[i, j] = - 1. / (τ + dt) * (rxx[i, j] + dt  * (  p2P * vxx / dx + p2d * vyy / dy + 0.5 * rxx[i, j]))
        ryy[i, j] = - 1. / (τ + dt) * (ryy[i, j] + dt  * (  p2d * vxx / dx + p2P * vyy / dy + 0.5 * ryy[i, j]))
        rxy[i, j] = - 1. / (τ + dt) * (rxy[i, j] + dt  * (  p2S * vyx / dx + p2S * vxy / dy + 0.5 * rxy[i, j]))

        pxx[i, j] += 0.5 * dt * rxx[i, j]
        pyy[i, j] += 0.5 * dt * ryy[i, j]
        pxy[i, j] += 0.5 * dt * rxy[i, j]
    end
  end

  




  heatmap(vy[:, :],framestyle = :box, clim=(-clip, clip), aspect_ratio = :equal, xlabel = "X", ylabel = "Y", title = "Wave Propagation", color = :seismic, size = (900, 900))
  
  # micro_data[n] = vx[Nx ÷ 2, Ny ÷ 2]
end every 10


#plot(micro)
